core/block.ts

```diff
@@ -313,44 +313,34 @@ export class Block implements IASTNodeLocation, IDeletable {
    * @suppress {checkTypes}
    */
   dispose(healStack: boolean) {
-    if (this.isDeadOrDying()) {
-      return;
-    }
-    // Terminate onchange event calls.
-    if (this.onchangeWrapper_) {
-      this.workspace.removeChangeListener(this.onchangeWrapper_);
-    }
+    if (this.isDeadOrDying()) return;
 
     this.unplug(healStack);
     if (eventUtils.isEnabled()) {
+      // Constructing the delete event is costly. Only perform if necessary.
       eventUtils.fire(new (eventUtils.get(eventUtils.BLOCK_DELETE))(this));
     }
-    eventUtils.disable();
+    this.workspace.removeTopBlock(this);
+    this.disposeInternal();
+  }
+
+  protected disposeInternal() {
+    if (this.isDeadOrDying()) return;
+
+    if (this.onchangeWrapper_) {
+      this.workspace.removeChangeListener(this.onchangeWrapper_);
+    }
 
+    eventUtils.disable();
     try {
-      // This block is now at the top of the workspace.
-      // Remove this block from the workspace's list of top-most blocks.
-      this.workspace.removeTopBlock(this);
       this.workspace.removeTypedBlock(this);
-      // Remove from block database.
       this.workspace.removeBlockById(this.id);
       this.disposing = true;
 
-      // First, dispose of all my children.
-      for (let i = this.childBlocks_.length - 1; i >= 0; i--) {
-        this.childBlocks_[i].dispose(false);
-      }
-      // Then dispose of myself.
-      // Dispose of all inputs and their fields.
-      for (let i = 0, input; input = this.inputList[i]; i++) {
-        input.dispose();
-      }
+      [...this.childBlocks_].forEach((c) => c.disposeInternal());
```

Comment:
This used to use the original array and go backwards. Now it uses a copy and goes forward. Why the change?

Reply:
So we discussed yesterday, and I said the issue was that `setParent(null)` updates the `childBlocks_` of the parent block. So calling `dispose` on children would mutate the array we were looping over. We could either deal with this by looping backwards (as previous) or by looping over a copy that doesn't get mutated.

However! I checked it again after we chatted and since we're no longer calling `unplug` on every block, we're also not calling `setParent(null)` on every block. So the array isn't being mutated at all, and we can just loop over it directly =)